Milestone 1

arl opened the pull request:
milestone 1 elements are present :
- dialog with configuration options and a **open device** button 
- mainwindow with input & output visible
- read/write from serial port

arl committed:
read serial and write output

ConnectDialog:
	- set default value for device
SessionManager:
	- open port
	- connect serial::readeyRead signal to output manager

arl committed:
insert serial output as plain text

arl committed:
subclass QLineEdit, connect input line to serial

subclassed QLineEdit <- LineInput
simple input that intercepts Enter/Return keypresses + emit signal with
the data entered

connected LineInput to SessionManager, that sends to serial port

arl committed:
move cursor at end of output before inserting

without this, if user was clicking inside the output, the next text
coming from the serial port was inserted at cursor position, not
at the end

arl committed:
add line ending member variable to InputLine

arl committed:
rename doc comments

for consistency with header :
	@brief -> \brief
	@param -> \param

arl committed:
fix a copy-paste typo

arl committed:
close connection before open

arl committed:
output display fixes

selection:
	eventual selection remains, even if some output text is appended
	and/or scrollbar moved
format:
	extra newlines fixed

scrollbar:
	scrollbar always pushed to the bottom when text is appended

gvaldambrini commented on inputline.cpp:
```diff
@@ -0,0 +1,27 @@
+
+}
+
+void InputLine::keyPressEvent(QKeyEvent *e)
```
Ma come mai ad oggi c'è bisogno di derivare dal QLineEdit e reimplementare il keyPressEvent?
Il QLineEdit emette un segnale quando viene premuto il "return" (vedi http://doc.qt.io/qt-5/qlineedit.html#returnPressed), non ti bastava agganciarti a quello dalla main window e chiamare da lì la sendToSerial?

aurelien-rainone replied:
infatti ho avevo valutato questa soluzione ma il returnPressed non manda il testo contenuto nel QLineEdit, quindi se facevo cosi, avrei dovuto a

arl committed:
remove InputLine

arl committed:
fix problems discussed with Lorenzo and Gianni

- removed dependency between OutputMgr & SessionMgr
	- removed QTextEdit pointer from OutputMgr
	- "un-subclass" QlineEdit and handle edit validation in
MainWindow
	- deleted lineedit.{h,cpp} files
	- reformulated some comments
	- code reformatted
	- some variable renaming to stick with Develer coding style
          guidelines

aurelien-rainone commented:
Ecco, avrei dovuto 'squash'are i due commit e combinareli in uno commit solo.
Ma comunque, ecco sono le cose cambiate tra ieri : 
  - removed dependency between OutputMgr & SessionMgr
  - removed QTextEdit pointer from OutputMgr
  - "un-subclass" QlineEdit and handle returnPressed signal in the MainWindow
  - deleted lineedit.{h,cpp} files
  - reformulated some comments
  - code reformatted and some variable renaming in order to stick with Develer coding style guidelines

arl committed:
Q_ASSERT instead of qDebug for port_cfg value check

aurelien-rainone commented:
Ecco, avrei dovuto 'squash'are i due commit e combinareli in uno commit solo.
Ma comunque, ecco sono le cose cambiate tra ieri :

 - removed dependency between OutputMgr & SessionMgr
 - removed QTextEdit pointer from OutputMgr
 - "un-subclass" QlineEdit and handle returnPressed signal in the MainWindow
 - deleted lineedit.{h,cpp} files
 - reformulated some comments
 - code reformatted and some variable renaming in order to stick with Develer coding style guidelines
 - Q_ASSERT instead of qDebug to check if port_cfg settings cast is ok, as if it isn't, that would be a programming error, not a software/hardware/user one

arl committed:
don't try to send anything if session is not open

also removed some qDebug

lmancini commented on mainwindow.cpp:
```diff
@@ -10,9 +10,15 @@
 #include "sessionmanager.h"
+#include "outputmanager.h"
+
+#include <QTextEdit>
```
Questo non dovrebbe servirti, non usi `QTextEdit` da nessuna parte in questo modulo.

lmancini commented on mainwindow.cpp:
```diff
@@ -37,3 +54,30 @@ void MainWindow::openConnectionDialog()
+
+    // insert data at end of 'edit' (but this clears any selection)
+    ui->textOutput->moveCursor(QTextCursor::End);
+    ui->textOutput->insertPlainText (textdata);
```
Coding style (spazio tra nome della funzione e parentesi).

lmancini commented on outputmanager.cpp:
```diff
@@ -16,3 +16,11 @@ OutputManager::OutputManager(QObject *parent) : QObject(parent)
+    buffer.append(data);
+
+    // notify that we have new data (in display format)
+    emit dataConverted(QString::fromLocal8Bit(data).remove('\n'));
```
Ho due dubbi su questa riga.

Il primo è che `remove` toglie tutte le occorrenze della stringa specificata come argomento (http://doc.qt.io/qt-5/qstring.html#remove-2), ed il payload che arriva a questa funzione è stato recuperato tramite `readAll`, che torna tutti i dati in attesa di essere letti (http://doc.qt.io/qt-5/qiodevice.html#readAll).  Se quindi in attesa c'erano due linee, ciascuna delle quali con il proprio a capo alla fine, questa riga li rimuove entrambi e fa sì che in output tu mostri una singola riga collassata, es.

Input:

```
ciao
mondo
```

Output:

```
ciaomondo
```

Mi sembra che potenzialmente tu possa perdere informazione così facendo; magari però nella pratica succede che i dati vengono resi disponibili da seriale solo quando arriva un terminatore di linea.  Volevo capire se avevi considerato questo problema.

Il secondo è che qua assumi che il terminatore che ti arriva sia `\n`, ma quello che mandi tu in scrittura su seriale è `\r\n`.  In base a cosa hai fatto questa distinzione?

arl replied:
At first i didn't realize that the line endings i was sending were also sent back by the serial port. So as i was sending `\r\n` as line endings, i was also receiving them. 

I didn't have the same output as the original cutecom. So to solve the _problem_, (that in fact i created myself by sending  `\r\n` instead of just  `\n`: unix endlines), i decided to remove all  `\n`  of the returned string.
The result on the output was ok, because QTextEdit interprets `\r\n`, `\r` and `\n` all in the same way.

Honestamento non ci avevo pensato tanto tempo. Ora faccio un push dove cambio LINE_ENDING, lo metto a `"\n"`. Vedremo piu tardi se puo servire a l'utente di cambiarla (come su cutecom)
E tolgo ovviamente il `remove('\n')`

lmancini commented on sessionmanager.cpp:
```diff
@@ -4,31 +4,93 @@
+    cfg_ok &= ok;
+
+    // a conversion didn't make it
+    Q_ASSERT(cfg_ok);
```
Ok, ti ricordo solo se vuoi usarla che la funzione che ti dicevamo ieri è `Q_ASSERT_X` (http://doc.qt.io/qt-5/qtglobal.html#Q_ASSERT_X).

arl committed:
remove unused include, fix coding style

arl committed:
correct line endings

arl committed:
Q_ASSERT_X

lmancini commented:
Lato codice mi sembra tutto ok, incontriamoci oggi per provarlo insieme sulle schede che stai usando.
